src/protobuf/generator.cr

```diff
@@ -243,9 +245,16 @@ end
 
 module Protobuf
   class Generator
-    # Turns `foo/bar/batz.proto` into `foo_bar_batz.pb.cr`
-    def self.output_filename(path)
-      path.split(".")[0..-2].join("").gsub(File::SEPARATOR, "_") + ".pb.cr"
+    def self.output_filename(input_path)
+      if ENV["PROTOC_GEN_CRYSTAL_PATHS"]? == "source_relative"
+        # Turns `foo/bar/batz.proto` into `foo/bar/batz.pb.cr`
+        output_path = input_path.split(".")[0..-2].join("") + ".pb.cr"
+        FileUtils.mkdir_p(File.dirname(output_path))
+        output_path
+      else
+        # Turns `foo/bar/batz.proto` into `foo_bar_batz.pb.cr`
+        input_path.split(".")[0..-2].join("").gsub(File::SEPARATOR, "_") + ".pb.cr"
+      end
```

Comment:
Note:
It would be cleaner to toggle this with a protoc-option like
protoc-gen-go does, i.e.: `--crystal_opt=paths=source_relative`

But the generator doesn't seem to support such options, yet,
so I went for an env var instead.